llvm/lib/Transforms/Vectorize/VPlanAnalysis.cpp

```diff
@@ -272,6 +272,14 @@ Type *VPTypeAnalysis::inferScalarType(const VPValue *V) {
             // TODO: Use info from interleave group.
             return V->getUnderlyingValue()->getType();
           })
+          .Case<VPExtendedReductionRecipe>(
+              [](const VPExtendedReductionRecipe *R) {
+                return R->getResultType();
+              })
+          .Case<VPMulAccumulateReductionRecipe>(
+              [](const VPMulAccumulateReductionRecipe *R) {
+                return R->getResultType();
+              })
```

Comment:
I think something like this should work:

```suggestion
          .Case<VPExtendedReductionRecipe, VPMulAccumulateReductionRecipe>(
              [](const VPExtendedReductionRecipe *R) {
                return R->getResultType();
              })
```

Reply:
The only common super class for them is `VPReductionRecipe`, so the lambda parameter would need to be cast. In which case it's probably cleaner to keep the structure as it is.

Reply:
The lambda should have used `const auto *R` (updated the suggestion) which is something that should be supported by the type switch, I think we use that feature in a number of places